### PR TITLE
refactor: remove `compileExponentiateExpr`

### PR DESCRIPTION
**Pull request by jaschdoc**
This is almost ready to be merged. I just have one question, see annotation below.

Related to https://github.com/flix/flix/issues/5713

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala [hunk 1]
```diff
@@ -68,6 +68,70 @@ object GenExpression {
...
 
+      case Float32Op.Exp =>
+        compileExpression(exp1, visitor, currentClass, lenv0, entryPoint)
+        visitor.visitInsn(F2D) // Sign extend to double
+        compileExpression(exp2, visitor, currentClass, lenv0, entryPoint)
+        visitor.visitInsn(F2D) // Sign extend to double
+        visitor.visitMethodInsn(INVOKESTATIC, JvmName.Math.toInternalName, "pow",
+          AsmOps.getMethodDescriptor(List(JvmType.PrimDouble, JvmType.PrimDouble), JvmType.PrimDouble), false)
+        visitor.visitInsn(D2F)
+        visitor.visitInsn(NOP) // TODO: Question: Does this not add any instruction to the visitor or are there invariants to maintain on the JVM??? Consider removing all NOPs
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala:79]: Also, see this comment:

> TODO: Question: Does this not add any instruction to the visitor or are there invariants to maintain on the JVM??? Consider removing all NOPs

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala [hunk 2]
```diff
@@ -68,6 +68,70 @@ object GenExpression {
...
+        visitor.visitInsn(L2D)
+        compileExpression(exp2, visitor, currentClass, lenv0, entryPoint)
+        visitor.visitInsn(L2D)
+        visitor.visitMethodInsn(INVOKESTATIC, JvmName.Math.toInternalName, "pow",
+          AsmOps.getMethodDescriptor(List(JvmType.PrimDouble, JvmType.PrimDouble), JvmType.PrimDouble), false)
+        visitor.visitInsn(D2L)
+        visitor.visitInsn(NOP)
+
+      // TODO: The following two cases are not handled by compileExponentiateExpr? Why is this the case? Are these impossible after typing? I'd argue we should handle them anyway.
+      case BigIntOp.Exp | BigDecimalOp.Exp => throw InternalCompilerException(s"Unexpected semantic operator: $sop.", exp1.loc) // compileExponentiateExpr(exp1, exp2, currentClass, visitor, lenv0, entryPoint, sop)
```
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala:133]: I don't think we know. Can we not just handle them like the rest?

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala (hunk 1 above)
Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala:79]: I'll compare it with some simple Java code :smile:

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala:133]: Yes, I'll get on it

**Comment by jaschdoc** [comment]
I'm not sure semantic operators `BigInt.Exp` and `BigDecimal.Exp` even make sense in Flix, so I'd argue they should be removed. The `Exp[a]` type class requires the type of `exp` to be `a -> a -> a` which doesn't match `BigInteger` and `BigDecimal` in Java. The `pow` methods have type `BigInteger / BigDecimal -> Int -> BigInteger / BigDecimal` i.e. not compatible with `a -> a -> a`. Maybe we should just implement it as a call in the library to the java function instead of implementing the `Exp` type class. Besides https://github.com/flix/flix/issues/5672 exists.

**Pull request merged**